service/worker/pernamespaceworker.go

```diff
@@ -258,83 +259,83 @@ func (wm *perNamespaceWorkerManager) getWorkerByNamespace(ns *namespace.Namespac
+	if count < 0 {
+		return workerAllocation{}, errInvalidConfiguration
+	} else if count == 0 {
+		return workerAllocation{0, 0}, nil
 	}
-	if desiredWorkersCount == 0 {
-		return &workerAllocation{0, 0}, nil
-	}
-	localCount, err := wm.getLocallyDesiredWorkersCount(ns, desiredWorkersCount)
+	localCount, err := w.getLocallyDesiredWorkers(count)
 	if err != nil {
-		return nil, err
+		return workerAllocation{}, err
 	}
-	return &workerAllocation{desiredWorkersCount, localCount}, nil
+	return workerAllocation{w.count, localCount}, nil
 }
 
-func (wm *perNamespaceWorkerManager) getConfiguredWorkersCountFor(ns *namespace.Namespace) (int, error) {
-	totalWorkers := wm.config.PerNamespaceWorkerCount(ns.Name().String())
-	if totalWorkers < 0 {
-		err := fmt.Errorf("%w namespace %s, workers count %d", errInvalidConfiguration, ns.Name(), totalWorkers)
-		return 0, err
-	}
-	return totalWorkers, nil
-}
-
-func (wm *perNamespaceWorkerManager) getLocallyDesiredWorkersCount(ns *namespace.Namespace, desiredNumberOfWorkers int) (int, error) {
-	key := ns.ID().String()
-	availableHosts := wm.serviceResolver.LookupN(key, desiredNumberOfWorkers)
+func (w *perNamespaceWorker) getLocallyDesiredWorkers(count int) (int, error) {
+	key := w.ns.ID().String()
+	availableHosts := w.wm.serviceResolver.LookupN(key, count)
 	hostsCount := len(availableHosts)
 	if hostsCount == 0 {
 		return 0, membership.ErrInsufficientHosts
 	}
-	maxWorkersPerHost := desiredNumberOfWorkers/hostsCount + 1
-	desiredDistribution := util.RepeatSlice(availableHosts, maxWorkersPerHost)[:desiredNumberOfWorkers]
+	maxWorkersPerHost := count/hostsCount + 1
+	desiredDistribution := util.RepeatSlice(availableHosts, maxWorkersPerHost)[:count]
 
-	isLocal := func(info membership.HostInfo) bool { return info.Identity() == wm.self.Identity() }
+	isLocal := func(info membership.HostInfo) bool { return info.Identity() == w.wm.self.Identity() }
 	result := len(util.FilterSlice(desiredDistribution, isLocal))
 	return result, nil
 }
 
-// called on namespace state change callback
-func (w *perNamespaceWorker) refreshWithNewNamespace(ns *namespace.Namespace, deleted bool) {
+func (w *perNamespaceWorker) setWorkerCount(count int) {
+	w.update(nil, false, &count, nil)
+}
+
+func (w *perNamespaceWorker) setWorkerOptions(opts sdkworker.Options) {
+	w.update(nil, false, nil, &opts)
+}
+
+// called on namespace state change callback, membership change, and dynamic config change
+func (w *perNamespaceWorker) update(ns *namespace.Namespace, nsDeleted bool, newCount *int, newOpts *sdkworker.Options) {
 	w.lock.Lock()
-	w.ns = ns
-	// namespace name can change, but don't update w.logger, otherwise we'd have to hold w.lock
-	// just to log.
+
+	if ns != nil {
+		w.ns = ns
+		// The name inside of *ns, which was used to initialize the logger, can change, but
+		// don't update w.logger here, otherwise we'd have to hold w.lock just to log.
```

Comment:
Could we hold a pointer to the active logger in `perNamespaceWorker` and swap that after initializing a new logger?

Reply:
Yeah, it's possible, I just meant it would be pretty annoying. Namespace renames are rare (I think it's not even exposed, it only happens as a step in namespace deletion), and I don't think it matters if we log the old name here.